internal/core/data/v2/handler/event.go

```diff
@@ -0,0 +1,82 @@
+//
+// Copyright (C) 2020 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package handler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	dataContainer "github.com/edgexfoundry/edgex-go/internal/core/data/container"
+	v2DataContainer "github.com/edgexfoundry/edgex-go/internal/core/data/v2/bootstrap/container"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/correlation"
+
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/container"
+	"github.com/edgexfoundry/go-mod-bootstrap/di"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+	model "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+
+	msgTypes "github.com/edgexfoundry/go-mod-messaging/pkg/types"
+)
+
+// The AddEvent function accepts the new event model from the controller functions
+// and invokes addEvent function in the infrastructure layer
+func AddEvent(e model.Event, ctx context.Context, dic *di.Container) (string, error) {
+	configuration := dataContainer.ConfigurationFrom(dic.Get)
+	dbClient := v2DataContainer.DBClientFrom(dic.Get)
+
+	err := checkDevice(e.Device, ctx, dic)
+	if err != nil {
+		return "", err
+	}
+
+	// Add the event and readings to the database
+	if configuration.Writable.PersistData {
+		addedEvent, err := dbClient.AddEvent(e)
+		if err != nil {
+			return "", err
+		}
+		e = addedEvent
+	}
+
+	putEventOnQueue(e, ctx, dic) // Push event to message bus for App Services to consume
+
+	return e.Id, nil
+}
+
+// Put event on the message queue to be processed by the rules engine
+func putEventOnQueue(evt model.Event, ctx context.Context, dic *di.Container) {
+	lc := container.LoggingClientFrom(dic.Get)
+	msgClient := dataContainer.MessagingClientFrom(dic.Get)
+	configuration := dataContainer.ConfigurationFrom(dic.Get)
+
+	lc.Info("Putting event on message queue")
+
+	evt.CorrelationId = correlation.FromContext(ctx)
+	var data []byte
+	var err error
+	// Re-marshal JSON content into bytes.
+	if clients.FromContext(ctx, clients.ContentType) == clients.ContentTypeJSON {
+		data, err = json.Marshal(evt)
+		if err != nil {
+			lc.Error(fmt.Sprintf("error marshaling event: %+v", evt))
+			return
+		}
+	}
+
+	msgEnvelope := msgTypes.NewMessageEnvelope(data, ctx)
+	err = msgClient.Publish(msgEnvelope, configuration.MessageQueue.Topic)
```

Comment:
It was agreed in the Core WG that all DTOs will have a version property and that V2 will publish the Event DTO to the message bus rather than the Event model. 

This impacts when the DTO is converted to a model, which should be just before storing it so the DTO is still available to publish to the message bus.